contributors/design-proposals/storage/local-storage-pv.md

```diff
@@ -761,3 +761,208 @@ The central manager will poll each worker for their discovered volumes and creat
+
+For LVM, the proposed workflow is:
+
+1. Check if there is already a PV pre-bound to the PVC, if so, return directly
+(This can happen when a volume has already been provisioned for a PVC,
+but has not been bound to the PVC by the controller yet).
+2. Check if the its capacity has been initialized in the StorageClass.
+If not, return error.
+3. Check if the class is available for dynamic provisioning (i.e. has an input of `volumeGroup`).
+If not, return error.
+4. Create `logical volume (LV)` object in the VG.
+5. Create a PV API object that is pre-bound to the PVC, and the path
+of local storage in the PV should be the path of the LV created above.
+
+**Note:** As it describes in [Volume Topology-aware Scheduling](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/volume-topology-scheduling.md),
+in whichever step it fails, the provisioner will signal back to the scheduler
+to retry dynamic provisioning by removing the "annSelectedNode" annotation of
+the PVC.
+
+For the purpose, it requires the permission to update PersistentVolumeClaims.
+
+##### Cleanup after Release
+
+Currently we already have a component `deleter` to cleanup volumes/PVs of static local volume,
+detail cleanup actions are managed via custom scripts/binaries that specified in `blockCleanerCommand`.
+
+Existing script `quick_reset.sh` calls `mkfs -F` and `wipefs` to cleanup a block.
+In addition to that, a hook named `volumeDeleteFunc` will be added into `deleter`,
+and for dynamic privisioned volumes, the deleting process should be:
+1. Call scripts/binaries in `blockCleanerCommand` to clean up data in the volume.
+2. Call `volumeDeleteFunc` to recycle the volume behind the PV (e.g. delete the LV for lvm).
+3. Delete the PV object.
+
+##### Expansion of Local Volume Plugin
+
+Currently, the path in `LocalVolumeSource` must be a directory. Thus, for dynamic
+provisioned volumes, the provisioner need to mount the LV to a host path first,
+to allow kubelet for another mount. It increases the complexity of the plugin.
+
+Therefore, in-tree local storage plugin will be updated to support both directory
+and block as a source of mount. The mount process would be:
+1. Detect if given path in volumeSource is mounted or not, if mounted,
+continue to work in the existing way.
+2. If not, Check if global path `{pluginDir}/mounts/{volumeID}` exists.
+and create the directory if not.
+3. Check if the global path is of mount point, if not, mount given path to
+the global path.
+4. Manage `mkfs` for the global path if it is newly created or mounted.
+5. Mount the global path to the path in pods.
+
+New field `FSType` will be added in `LocalVolumeSource` to support `mkfs` in the mounter.
+
+```
+type LocalVolumeSource struct {
+	// The full path to the volume on the node
+	// For alpha, this path must be a directory
+	// Once block as a source is supported, then this path can point to a block device
+	Path string

```

Comment:
I think the confusing part will be that there are 3 different behaviors for path depending on the underlying type:

1. Volume mode = FS, path is a FS.  We bind mount the path to pod directory
2. Volume mode = FS, path is a device. We mkfs if needed, and globally mount the device.  Then bind mount global path to pod directory.
3. Volume mode = block, path is a device. Passthrough.

I wonder if we could make this behavior more explicit if we supported true "fs-on-block" volume mode binding (ie allow PVC volumeMode fs to bind to PV volumeMode block if the plugin supports it).  I know it was initially rejected for simplicity, but it really does make the local volume behavior a bit confusing.

@jsafrane would appreciate your thoughts as you originally had strong objections to the overloading of the Path field in the initial design.

Reply:
Yeah, listen to me :-). It *is* confusing, however it's for separate PR on block storage.

With regard to local storage, I expect that "path is a FS" should be used in very special cases where underlying platform can't provide us block devices.

"path is a device" is what we should message to users, with a small note somewhere below that it might be a directory if it's really needed.

Reply:
Thanks for your input. So IIUC, we can propose the volumeMode binding change separately, and go with this approach for now to be unblocked.  It also means for static discovery, we need a new parameter to indicate if block device should be volumeMode fs or block.

I really want to go with the explicit volumeModes approach before we make this behavior beta though.  Otherwise we'll still have to support both ways.

Reply:
So for now, besides `Path` and `FsType`, we'll need another field like `SourceType` in the LocalVolumeSource, to indicate its type, is it?

Another notable thing it that currently interface UnmountDevice looks like this:
```
UnmountDevice(deviceMountPath string) error
```
if only the interface is implemented, unmount will be managed for sure, but if the source path is of fs, we don't need global mount/unmount indeed.
So maybe we should add extra parameter to the interface for check?

Reply:
or in the implementation `UnmountDevice` for local storage, we can have a internal agreement on the format of global mount path, and basing on the format of given deviceMountPath, we can decide whether global unmount is needed or not.

Reply:
I don't think we need a `SourceType` in the LocalVolumeSource.  I meant we need a new field in `MountConfig` of static provisioner. Right now, it assumes if the device is a block device, then PV.volumeMode is Block.

I wouldn't change the `UnmountDevice` interface, we can just check internally.